### PR TITLE
Feature/add topic

### PR DESCRIPTION
**Pull request by brunoparma88**
create topics index

### CODE REVIEW

File: app/controllers/api/v1/topics_controller.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module Api
+  module V1
+    class TopicsController < Api::V1::ApiController
+      def index
+        @topics = Topic.get_all
```
Comment (jpascual1994) [inline app/controllers/api/v1/topics_controller.rb]: Do we need to use this scope?
```suggestion
        @topics = Topic.all
```

File: app/models/topic.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+#
+#  id         :bigint           not null, primary key
+#  name       :string           not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+class Topic < ApplicationRecord
+  validates :name, presence: true
+  has_one_base64_attached :avatar
+  scope :get_all, -> { limit(50) }
```
Comment (jpascual1994) [inline app/models/topic.rb]: maybe we can remove this scope as we don't need to limit the number of topics that we return

File: spec/request/api/v1/topic/index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  let(:filename) { 'generic-logo.png' }
+  let(:file) do
+    {
+      io: File.open(
+        File.expand_path(File.join(File.dirname(__FILE__) + '/../../../..', 'fixtures', filename))
+      ),
+      filename: filename
+    }
+  end
```
Comment (jpascual1994) [inline spec/request/api/v1/topic/index_spec.rb]: Maybe we can create a trait on the factory so we always create a topic with the same image

File: spec/request/api/v1/topic/index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  it 'returns success' do
+    expect(subject).to have_http_status(:success)
+  end
+
+  it 'assigns the specified filename' do
+    topic.avatar.attach(file)
+    topic.save!
+    expect(topic.avatar.attached?).to be
+  end
```
Comment (jpascual1994) [inline spec/request/api/v1/topic/index_spec.rb]: We should create a model test for topics and move this check to that file

File: Gemfile [hunk 5]
```diff
@@ -51,9 +51,11 @@ gem 'webpacker', '~> 5.0'
 # Turbolinks makes navigating your web application faster. Read more: https://github.crailrom/turbolinks/turbolinks
 gem 'turbolinks', '~> 5'
 # Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
+gem 'active_storage_base64'
```
Comment (martinjaimem) [inline Gemfile]: we should add a fixed version, you can check the one installed in the `Gemfile.lock`

File: app/controllers/api/v1/api_controller.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      rescue_from ActiveRecord::RecordInvalid,         with: :render_record_invalid
+      rescue_from ActionController::RoutingError,      with: :render_not_found
+      rescue_from AbstractController::ActionNotFound,  with: :render_not_found
+
+      def status
+        render json: { online: true }
+      end
+
+      def render_not_found(exception)
+        logger.info(exception) # for logging
```
Comment (martinjaimem) [inline app/controllers/api/v1/api_controller.rb]:
```suggestion
        logger.info { exception }
```

slightly more performant

File: app/models/topic.rb (hunk 2 above)
Comment (martinjaimem) [reply app/models/topic.rb]: agree, but we should paginate them

File: spec/models/topic_spec.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+describe Topic do
+  describe 'image' do
```
Comment (martinjaimem) [inline spec/models/topic_spec.rb:17]: what about testing the validations?

File: app/models/topic.rb (hunk 2 above)
Comment (sebaherrera07) [reply app/models/topic.rb]: Probably it is better to do the pagination in a separate PR, so you can just remove this scope now.

File: app/controllers/api/v1/api_controller.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      respond_to :json
+
+      rescue_from ActiveRecord::RecordNotFound,        with: :render_not_found
+      rescue_from ActiveRecord::RecordInvalid,         with: :render_record_invalid
+      rescue_from ActionController::RoutingError,      with: :render_not_found
+      rescue_from AbstractController::ActionNotFound,  with: :render_not_found
+
+      def status
+        render json: { online: true }
+      end
```
Comment (martinjaimem) [inline app/controllers/api/v1/api_controller.rb:19]: what about making a small test for this?

File: db/migrate/20210705182718_create_topics.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+class CreateTopics < ActiveRecord::Migration[6.1]
+  def change
+    create_table :topics do |t|
+      t.string :name, null: false
```
Comment (sebaherrera07) [inline db/migrate/20210705182718_create_topics.rb:4]: `t.string :name, null: false, index: true`

When you think you will use a column for ordering the records, or for doing searches, it is good to consider adding a database index to it.

Be careful and try not to abuse when a model has many columns, to add index to every column or things like that.

File: spec/request/api/v1/topic/index_spec.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
```
Comment (martinjaimem) [inline spec/request/api/v1/topic/index_spec.rb]:
```suggestion
```

not needed 👍

File: spec/models/topic_spec.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
```
Comment (martinjaimem) [inline spec/models/topic_spec.rb:15]:
```suggestion
```

Not needed

File: config/routes.rb [hunk 12]
```diff
@@ -10,6 +10,7 @@
   namespace :api do
     namespace :v1, defaults: { format: :json } do
       devise_scope :user do
+        resources :topics, only: :index
```
Comment (martinjaimem) [inline config/routes.rb:14]: can we test this routing?

File: config/initializers/rack_cors.rb [hunk 13]
```diff
@@ -0,0 +1,13 @@
+module App
+  class Application < Rails::Application
+    config.middleware.insert_before 0, Rack::Cors do
+      allow do
+        origins '*'
+        resource '*',
+                 headers: :any,
+                 methods: %i[get post options put delete],
+                 expose: %w[access-token uid client]
```
Comment (sebaherrera07) [inline config/initializers/rack_cors.rb]: Just curious, why are this changes needed?

File: spec/request/api/v1/topic/index_spec.rb [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+describe 'GET api/v1/topics', type: :request do
+  subject do
+    get api_v1_topics_path, headers: auth_headers, as: :json
+    response
+  end
+
+  let(:user)    { create :user }
```
Comment (martinjaimem) [inline spec/request/api/v1/topic/index_spec.rb]:
```suggestion
  let(:user)    { create(:user) }
```

File: spec/request/api/v1/topic/index_spec.rb [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+  let!(:topics) { create_list :topic, 5 }
```
Comment (martinjaimem) [inline spec/request/api/v1/topic/index_spec.rb]:
```suggestion
  let!(:topics) { create_list(:topic, 5) }
```

File: config/application.rb [hunk 16]
```diff
@@ -10,6 +11,7 @@ module Socialnetwork
   class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 6.1
+    config.active_storage.service = :local
```
Comment (sebaherrera07) [inline config/application.rb]: I'm not sure if you should have the configuration at the environment level,
Like some value for `development.rb` and other values for `production.rb` and `test.rb`

Eventually you can look at the `rails_api_base` as a guide.

File: spec/models/topic_spec.rb [hunk 17]
```diff
@@ -0,0 +1,10 @@
...
+describe Topic do
+  describe 'image' do
+    let(:topic) { create(:topic, :with_avatar) }
+    it 'have avatar attached' do
+      expect(topic.avatar.attached?).to be
+    end
+  end
```
Comment (sebaherrera07) [inline spec/models/topic_spec.rb:22]: You should also add tests for the model validations, like `presence: true` or `uniqueness: true`
Check this gem that might be useful:
https://github.com/thoughtbot/shoulda-matchers

File: app/models/topic.rb [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+#
+# Table name: topics
+#
+#  id         :bigint           not null, primary key
+#  name       :string           not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+class Topic < ApplicationRecord
+  validates :name, presence: true
```
Comment (sebaherrera07) [inline app/models/topic.rb]: Eventually you could also add a `uniqueness: true` validation, so that you don't have more than 1 topic with the same `name`.

And remember to add model specs for validations.

File: app/models/application_record.rb [hunk 19]
```diff
@@ -1,5 +1,6 @@
 # frozen_string_literal: true
 
 class ApplicationRecord < ActiveRecord::Base
+  include ActiveStorageSupport::SupportForBase64
```
Comment (sebaherrera07) [inline app/models/application_record.rb:4]: I'm not sure if its the best to place this line here or just do it in the models that will use active record.

Because if you place it here, the models that don't need active record would still have the methods that this line adds.

Try to see examples of where other people put this line.

**Review by sebaherrera07 — COMMENTED** [review]
Looks good!! 👏 

I left some small comments.

File: config/initializers/rack_cors.rb (hunk 13 above)
Comment (brunoparma88) [reply config/initializers/rack_cors.rb]: i dont need it!

File: config/application.rb (hunk 16 above)
Comment (brunoparma88) [reply config/application.rb]: Yes, I had it configured by environments

File: app/models/application_record.rb (hunk 19 above)
Comment (brunoparma88) [reply app/models/application_record.rb:4]: I was investigating and I saw that all users add them there. 
It is also recommended to apply it there on the official website

https://github.com/rootstrap/active-storage-base64/blob/master/README.md

File: app/controllers/api/v1/api_controller.rb [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+      def render_not_found(exception)
+        logger.info  { exception } # for logging
```
Comment (sebaherrera07) [inline app/controllers/api/v1/api_controller.rb]: No need for the comment

File: app/controllers/api/v1/api_controller.rb [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+        render json: { error: "Couldn't find the record" }, status: :not_found
+      end
+
+      def render_record_invalid(exception)
+        logger.info(exception) # for logging
```
Comment (sebaherrera07) [inline app/controllers/api/v1/api_controller.rb:27]: You missed changing this to 
```
logger.info { exception }
```

File: app/models/topic.rb [hunk 22]
```diff
@@ -0,0 +1,18 @@
+# == Schema Information
+#
+# Table name: topics
+#
+#  id         :bigint           not null, primary key
+#  name       :string           not null
```
Comment (sebaherrera07) [inline app/models/topic.rb:6]: You should run the migrations again so that the annotate comments include the index here

**Review by sebaherrera07 — APPROVED** [review]
💪 Nice!
There are still some comments to resolve, but I think none of them are big.

**Review by jpascual1994 — APPROVED** [review]
